[Silas Yeo] iP

comicalromance opened the pull request:
# DukePro
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2223s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 🐱

And it is **FREE**!

## Features:
 
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the main method:

```java
public class Duke {
    public static void main(String[] args) {
        new Duke(FILE_LOCATION, FOLDER_LOCATION).run();
    }
}
```

Commit:
Add Gradle support

comicalromance committed:
Level-1

comicalromance committed:
Level-2

comicalromance committed:
Level-3

comicalromance committed:
Level-4

comicalromance committed:
A-TextUiTesting

comicalromance committed:
Level-5

comicalromance committed:
Level-6

comicalromance committed:
A-Enums

comicalromance committed:
Add save and load functionality (not complete)

aroldoraegan committed:
Finish Adding save and load functionality

aroldoraegan committed:
Change save file location

aroldoraegan committed:
Add date time functionality

aroldoraegan committed:
Merge branch 'branch-Level-8'

aroldoraegan committed:
Fix load function with date format

Also reimplemented bye

comicalromance committed:
Add OOP

did the OOP

comicalromance committed:
Add packaging

comicalromance committed:
Merge branch 'add-gradle-support'

comicalromance committed:
Add JUnit tests

comicalromance committed:
Modify location and gradle class

comicalromance committed:
Add JavaDoc for main classes

comicalromance committed:
Revert "Add JavaDoc for main classes"

This reverts commit ba7ef1a1b385432ac3a9ccc8787b19c6734cbb13.

comicalromance committed:
Modify code to adhere to standards

comicalromance committed:
Add Find command

comicalromance committed:
Revert "Revert "Add JavaDoc for main classes""

This reverts commit b7dff63b1b4176f8b9500acbdbb430a45453a87f.

comicalromance committed:
Merge branch 'branch-Level-9'

comicalromance committed:
Merge branch 'branch-A-CodingStandard'

comicalromance committed:
Update FindCommand

Update to use getter for retrieving TaskList

tingkai-mai commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,15 @@
+
+    public abstract void execute(TaskList tasks, Ui ui, Storage storage);
+
+    public boolean isClose() {
```
I think the naming "isClose" might be a bit confusing, how about something like "isExit"?

comicalromance replied:
I agree that exit might cause less ambiguity for users!

tingkai-mai commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,20 @@
+
+public class DeleteCommand extends Command {
+
+    int index;
```
Hmm, do you think this should be a private variable?

comicalromance replied:
Good spot! I should specify the type of access modifier explicitly instead of using the default and making people wonder if it is deliberately or accidentally default.

tingkai-mai commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,39 @@
+            isDone = false;
+    }
+
+    public String getStatusIcon() {
```
Great job separating this into its own function! This makes checking for "X" very clear.

comicalromance replied:
Thanks for the compliment!

tingkai-mai commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,39 @@
+package duke.task;
+
+public class Task {
+    enum Tag {
```
The enum values for the Tags aren't immediately clear, especially when they're used within the code. (e.g. I wasn't too sure what Tag.T meant at first.) Might you want to consider renaming this?

comicalromance replied:
I understand that the tags may be a bit confusing, but this method is used to reduce overhead when calling the toString method which simply prints tag. However, this definitely can be resolved by modifying the tag names and adding a new method to return a one char tag for printing.

tingkai-mai commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,112 @@
+     * @return Command to be executed.
+     * @throws Exception  If there is an error in parsing.
+     */
+    public static Command parse(String input) throws Exception {
```
Is it better to throw a DukeException here?

comicalromance replied:
While it would be great to throw a DukeException for consistency, some parts of the code can throw other exceptions that I have not yet handled as an DukeException, so I can only throw Exception as a quick fix for now.

tingkai-mai commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,59 @@
+     * @param tasks A list of strings containing lines to save.
+     * @throws Exception if save file cannot be written.
+     */
+    public void save(TaskList tasks) throws Exception {
```
Since you threw a DukeException for `load`, you might want to consider throwing that for `save` as well!

comicalromance replied:
Good spot! Will handle the save function with DukeException in mind.

tingkai-mai reviewed (commented):
LGTM! I think I would suggest renaming some of the enum values to make it clearer, and perhaps throw/catch exceptions using DukeException rather than Exception.

comicalromance committed:
Revise code under review

Made modifications to several access modifiers, method names, enum values and exceptions returned.

comicalromance committed:
Add CheckStyle and Modify code accordingly

comicalromance committed:
Add GUI

rexong commented on src/main/java/duke/DialogBox.java:
```diff
@@ -0,0 +1,58 @@
+    @FXML
+    private Label dialog;
+    @FXML
+    private ImageView displayPicture;
```
Perhap it is better to name is as a noun? example `pictureDiplayer`.

rexong commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,55 @@
+        this(FILE_LOCATION, FOLDER_LOCATION);
+    }
+
+    public boolean getLoaded() {
```
Maybe the method name can be changed to `isLoaded` instead?

rexong commented on src/test/java/duke/ParserTest.java:
```diff
@@ -0,0 +1,28 @@
+package duke;
+
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.fail;
+
+import org.junit.jupiter.api.Test;
+
+import duke.command.EventCommand;
+import duke.command.ExitCommand;
+import duke.command.MarkCommand;
+import duke.command.UnmarkCommand;
```
I like how your import statements are in order. 👍

rexong reviewed (commented):
Your code looks good. Most of it is following the coding standard! Just some of the minor suggestions regarding your method and attributes naming.

tanhl2000 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,85 @@
+            if (duke.getLoaded()) {
+                print("File successfully loaded!");
+            } else {
+                print("Error parsing load file");
```
This can be put under Exception handling instead as it is an error message that should show with an exception. DukeException could also be expanded upon to have multiple more specific exceptions.

tanhl2000 reviewed (commented):
Overall great job with the code! Coding standards mostly followed as well.

comicalromance committed:
Update Main class for Gradle

mainClassName in the Gradle config is set to duke.Duke.

The current main class should be duke.Launcher instead.

The variable has thus been changed to duke.Launcher.

comicalromance committed:
Add assertions to important methods

Some methods were missing assertions that could be useful for
developers.

Assertions could be used to make sure unexpected behaviour does not
occur during runtime.

Let's add assert statements to two segments of code:
* In Parser, an assert in getTaskName was added to ensure the input
array has more than one element, indicating a name is present
* In  TaskList, an assert in the constructor was added to ensure that
if the task parsed is a ToDo (msg.length == 1), that no date has been
parsed as expected.

comicalromance committed:
Refactor parse method in Parser

Parser method is extensive with numerous duplicate code fragments.

Excessively long code is difficult to both read and debug.

Let's refactor code by extracting two methods:
* create parseDateTime to convert string into LocalDateTime
* create tokenizeInput to convert string into tokens with error
handling

comicalromance committed:
Merge pull request #1 from comicalromance/branch-A-Assertions

Add assertions to important methods

comicalromance committed:
Merge branch 'master' into branch-A-CodeQuality

comicalromance committed:
Merge pull request #2 from comicalromance/branch-A-CodeQuality

Refactor parse method in Parser

comicalromance committed:
Add sort functionality

There is no simple way to organize tasks in the task list.

Users who want to sort their tasks by upcoming deadline or alphabetical
order could not do so easily.

Let's add a sort command that allows users to sort the list either
alphabetically or chronologically, and in ascending or descending
order.

Let's also refactor the Task class to include time as well.

comicalromance committed:
Added JavaDoc comments

Let's add more JavaDoc comments to previously undocumented code.

comicalromance committed:
Update sort command

Sort command only parses user input if it strictly matches the listed parameters.
Users need to type entire parameter, which is long and not user-friendly.

Let's check if user input matches the starting characters of the parameter instead of the whole string.

comicalromance committed:
Update UI

Error messages are undifferentiated from normal Duke messages.
Users may be unsure when they are getting error messages.

Let's add an error message dialog to better distinguish the two.

comicalromance committed:
Update User Guide

Let's update the user guide for a more user-friendly experience.

comicalromance committed:
Remove Duplicate in UG

Remove duplicate entry for list in user guide.
